ledger/internal/apptxn_test.go

```diff
@@ -3058,3 +3058,222 @@ check:
 	txns(t, l, eval, &fundA, &callA)
 	endBlock(t, l, eval)
 }
+
+func TestForeignAppAccountsAccessible(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	genBalances, addrs, _ := ledgertesting.NewTestGenesis()
+	testConsensusRange(t, 32, 0, func(t *testing.T, ver int) {
+		dl := NewDoubleLedger(t, genBalances, consensusByNumber[ver])
+		defer dl.Close()
+
+		appA := txntest.Txn{
+			Type:   "appl",
+			Sender: addrs[0],
+			ApprovalProgram: main(`
+int 3
+int 3
+==
+assert
+`),
+		}
+
+		appB := txntest.Txn{
+			Type:   "appl",
+			Sender: addrs[0],
+			ApprovalProgram: main(`
+itxn_begin
+	int pay;                itxn_field TypeEnum
+	int 100;     		    itxn_field Amount
+	txn Applications 1
+	app_params_get AppAddress
+	assert
+	itxn_field Receiver
+itxn_submit
+`),
+		}
+
+		dl.beginBlock()
+		dl.txgroup("", &appA, &appB)
+		vb := dl.endBlock()
+		index0 := vb.Block().Payset[0].ApplicationID
+		index1 := vb.Block().Payset[1].ApplicationID
+
+		fund1 := txntest.Txn{
+			Type:     "pay",
+			Sender:   addrs[0],
+			Receiver: index1.Address(),
+			Amount:   1_000_000_000,
+		}
+		fund0 := fund1
+		fund0.Receiver = index0.Address()
+		fund1.Receiver = index1.Address()
+
+		callTx := txntest.Txn{
+			Type:          "appl",
+			Sender:        addrs[2],
+			ApplicationID: index1,
+			ForeignApps:   []basics.AppIndex{index0},
+		}
+
+		dl.beginBlock()
+		if ver <= 32 {
+			dl.txgroup("invalid Account reference", &fund0, &fund1, &callTx)
+			dl.endBlock()
+			return
+		}
+
+		dl.txgroup("", &fund0, &fund1, &callTx)
+		vb = dl.endBlock()
+
+		require.Equal(t, index0.Address(), vb.Block().Payset[2].EvalDelta.InnerTxns[0].Txn.Receiver)
+		require.Equal(t, uint64(100), vb.Block().Payset[2].EvalDelta.InnerTxns[0].Txn.Amount.Raw)
+	})
+}
+
+// While accounts of foreign apps are available in most context, they still
+// cannot be used as mutable references; ie the accounts cannot be used by
+// opcodes that modify local storage.
+func TestForeignAppAccountsImmutable(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	genBalances, addrs, _ := ledgertesting.NewTestGenesis()
+	l := newTestLedger(t, genBalances)
+	defer l.Close()
```

Comment:
It would be nice if this was also testConsensusRange so that it tests all the versions, starting with the first that works, through vFuture as we add more and more versions.